### PR TITLE
Translate all remaining R functionality

### LINKED ISSUE #2: implement rest of package

### PR DESCRIPTION
**Pull request by brettedw**
This PR is a comprehensive conversion of the CFFDRS (Canadian Forest Fire Danger Rating System) R package to Python.
- closes #2  

### Key Changes
- Implemented as scalar operations to match existing python code, without batch modes or vectorization

#### New Modules
- **Core Models** `cffdrs/models.py`: Dataclasses for FBP inputs and outputs (`FBPInput`, `FBPPrimaryOutput`, `FBPSecondaryOutput`, `FBPAllOutput`)
- All the rest of the modules match the R code in file and function names
- **Utilities**:
  - `cffdrs/constants.py` - Fuel type constants
  - `cffdrs/r_helpers.py` - R compatibility helpers

#### Test Suite
- Test coverage uses all of the same csv data from the R package
- Nearly all tests are passing, with the exception of `slope_calc.py`, `sdmc.py` and a single `fbp` test. These failing tests have reasoning in their `@pytest` decorators. 
- `slope_calc.py` test failures are edge case failures seemingly from some difference between R and Python mathematical operations, but shouldn't have a practical effect on it's outputs (all tests that use the `slope_adjustment` function pass). The failures all occur when `WSV` is very small, <1e-11 
    - If you want to see the test results, you can comment out the `@pytest.mark.xfail` and run the test  
- I've left a comment on the single failing `fbp` test, I believe it might be due to a small error in the R code/test data
- These failing tests are currently skipped in the test runner, all other tests run and pass

#### Configuration Updates
- `pyproject.toml`: Added `ruff` for consistent code formatting.

#### R compatibility
- Function names designed to match R package conventions for easy translation
- `r_helpers.py` provides R-style math operations for specific cases like div by 0

### CODE REVIEW

File: cffdrs/tests/test_fbp.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+
+############
+# fbp_08 and fbp_09 not included as the same data exists in other csv tests
+############
+
+
+@pytest.mark.xfail(
+    reason="I believe default CFL should be 0, not 1 as it's currently written - https://github.com/cffdrs/cffdrs_r/blob/4e40dd3af841f3a708abd83d7f2d43fefb08649a/R/fire_behaviour_prediction.r#L38"
+)
```
Comment (brettedw) [inline cffdrs/tests/test_fbp.py:183]: I think it's possible there's a minor mistake in the fbp implementation and test data from the R package, that only surfaces if no input is provided

File: cffdrs/buildup_effect.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    :param fuel_type: The Fire Behaviour Prediction FuelType
+    :param bui: The Buildup Index value
+
+    :returns: BE Build up effect
+    """
+    # Eq. 54 (FCFDG 1992) The Buildup Effect
+    fuel = FUEL_TYPE_DEFAULTS.get(fuel_type)
+
+    if fuel is None:
+        # R behavior: if fuel unknown and BUI > 0 -> NA, else -> 1
```
Comment (conbrad) [inline cffdrs/buildup_effect.py]: What do you think about capturing R lib behaviour/decisions in the doc comment? IDEs would be able to surface that easier for devs looking for documentation.

File: cffdrs/constants.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    "S1",
+    "S2",
+    "S3",
+    "O1A",
+    "O1B",
+    "NF",
+    "WA",
+]
+
+# Fuel type mappings
```
Comment (dgboss) [inline cffdrs/constants.py:30]: Consider adding a comment about what `BUIo` and `Q` represent for `FUEL_TYPE_DEFAULTS` and `a`, `b` and `c` for `FUEL_TYPE_ROS` if they have a useful meaning.

File: cffdrs/direction.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    Direction definer
+
+    :param bearingT1T2: Bearing between T1 and T2
+    :param bearingT1T3: Bearing between T1 and T3
+    :param ThetaAdeg: Direction
+
+    :returns: dir - a direction in degrees
+    """
+    dir = None
+    if bearingT1T2 > 0 and bearingT1T3 > 0 and bearingT1T2 > bearingT1T3:
```
Comment (dgboss) [inline cffdrs/direction.py:17]: Are all the `elif` conditions guaranteed to be mutually exclusive? I noticed that the R implementation assigns `DIR` multiple times.

File: cffdrs/direction.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        dir = bearingT1T2 + ThetaAdeg - 360
+    elif bearingT1T2 > 90 and bearingT1T3 < -90 and bearingT1T2 + ThetaAdeg < 180:
+        dir = bearingT1T2 + ThetaAdeg
+    elif bearingT1T2 < -90 and bearingT1T3 > 90 and bearingT1T2 - ThetaAdeg < -180:
+        dir = bearingT1T2 - ThetaAdeg + 360
+    elif bearingT1T2 < -90 and bearingT1T3 > 90 and bearingT1T2 - ThetaAdeg > -180:
+        dir = bearingT1T2 - ThetaAdeg
+    if dir is not None:
+        if dir < -180:
+            dir += 360
```
Comment (dgboss) [inline cffdrs/direction.py:39]: Are these `+ 360` and `-360` correct? The R implementation has `+ 10` and `-10`.

Comment (brettedw) [reply cffdrs/direction.py:39]: Good catch, you're right, this implementation is slightly different than the R implementation. I was a bit confused by the +/- 10 and in my conversations with a higher power it also thought the +/- 10 was a bug. However, a scenario where these implementations differ never came up in the tests for `pros`/`lros` where this function is used (there are no unit tests for `direction` in the R library).

I just generated some random inputs and the majority of them are the same:

Python outputs
``` 
Test 1: bearingT1T2=10, bearingT1T3=5, ThetaAdeg=20 -> dir=-10
Test 2: bearingT1T2=5, bearingT1T3=10, ThetaAdeg=20 -> dir=25
Test 3: bearingT1T2=-10, bearingT1T3=-5, ThetaAdeg=20 -> dir=10
Test 4: bearingT1T2=-5, bearingT1T3=-10, ThetaAdeg=20 -> dir=-25
Test 5: bearingT1T2=45, bearingT1T3=-45, ThetaAdeg=30 -> dir=15
Test 6: bearingT1T2=-45, bearingT1T3=45, ThetaAdeg=30 -> dir=-15
Test 7: bearingT1T2=100, bearingT1T3=-100, ThetaAdeg=50 -> dir=150
Test 8: bearingT1T2=100, bearingT1T3=-100, ThetaAdeg=90 -> dir=-170
Test 9: bearingT1T2=-100, bearingT1T3=100, ThetaAdeg=50 -> dir=-150
Test 10: bearingT1T2=-100, bearingT1T3=100, ThetaAdeg=90 -> dir=170
Test 11: bearingT1T2=10, bearingT1T3=5, ThetaAdeg=200 -> dir=170
Test 12: bearingT1T2=5, bearingT1T3=10, ThetaAdeg=200 -> dir=-155
Test 13: bearingT1T2=120, bearingT1T3=-120, ThetaAdeg=70 -> dir=-170
Test 14: bearingT1T2=-120, bearingT1T3=120, ThetaAdeg=70 -> dir=170
```

R outputs
```
Test 1: bearingT1T2=10.0, bearingT1T3=5.0, ThetaAdeg=20.0 -> DIR=-10.0
Test 2: bearingT1T2=5.0, bearingT1T3=10.0, ThetaAdeg=20.0 -> DIR=25.0
Test 3: bearingT1T2=-10.0, bearingT1T3=-5.0, ThetaAdeg=20.0 -> DIR=10.0
Test 4: bearingT1T2=-5.0, bearingT1T3=-10.0, ThetaAdeg=20.0 -> DIR=-25.0
Test 5: bearingT1T2=45.0, bearingT1T3=-45.0, ThetaAdeg=30.0 -> DIR=15.0
Test 6: bearingT1T2=-45.0, bearingT1T3=45.0, ThetaAdeg=30.0 -> DIR=-15.0
Test 7: bearingT1T2=100.0, bearingT1T3=-100.0, ThetaAdeg=50.0 -> DIR=150.0
Test 8: bearingT1T2=100.0, bearingT1T3=-100.0, ThetaAdeg=90.0 -> DIR=-170.0
Test 9: bearingT1T2=-100.0, bearingT1T3=100.0, ThetaAdeg=50.0 -> DIR=-150.0
Test 10: bearingT1T2=-100.0, bearingT1T3=100.0, ThetaAdeg=90.0 -> DIR=170.0
Test 11: bearingT1T2=10.0, bearingT1T3=5.0, ThetaAdeg=200.0 -> DIR=-180.0
Test 12: bearingT1T2=5.0, bearingT1T3=10.0, ThetaAdeg=200.0 -> DIR=195.0
Test 13: bearingT1T2=120.0, bearingT1T3=-120.0, ThetaAdeg=70.0 -> DIR=-170.0
Test 14: bearingT1T2=-120.0, bearingT1T3=120.0, ThetaAdeg=70.0 -> DIR=170.0
```

Test 11 & 12 is where they differ, but I actually think the Python implementation is correct. For test 12, the direction would be calculated as 205, then adjusted to 195, which I don't quite understand. 
Whereas in Python the output is -155 (equivalent to 205 on a 360-degree circle)

This may be a question for the CFS devs, but wanted to write all this out because I forgot to revisit it

File: cffdrs/constants.py (hunk 3 above)
Comment (brettedw) [reply cffdrs/constants.py:30]: Added in [be561d3](https://github.com/cffdrs/cffdrs_py/pull/13/commits/be561d3f4d066f4a81c1d484ddb58af53805b20c)

File: cffdrs/direction.py (hunk 5 above)
Comment (dgboss) [reply cffdrs/direction.py:39]: An adjustment of 10 degrees does seem odd. At first glance it would seem that you would want to normalize the calculated direction to fit within 0-360 but neither implementation does that.

File: cffdrs/direction.py (hunk 4 above)
Comment (brettedw) [reply cffdrs/direction.py:17]: They should be mutually exclusive and produce the same results (apart from the +/- 10 vs 360)

File: cffdrs/direction.py (hunk 5 above)
Comment (dgboss) [reply cffdrs/direction.py:39]: Looks like the original paper normalizes to -180 to 180, so your Python implementation is correct I'd say. Maybe you could include something in the doc string to clarify this?

Comment (brettedw) [reply cffdrs/direction.py:39]: Thanks for looking into the paper! Didn't realize it was in there. Good call, I'll put a note in

**Review by conbrad — APPROVED** [review]

**Review by dgboss — APPROVED** [review]
Strong work!

File: cffdrs/direction.py (hunk 5 above)
Comment (brettedw) [reply cffdrs/direction.py:39]: Docstring note added in [3fb0bd0](https://github.com/cffdrs/cffdrs_py/pull/13/commits/3fb0bd07c2096040056e49d3cfefb8e7fb788b5a)

**Comment by brettedw** [comment]
@BadgerOnABike @jordan-evens This is quite a large PR, but it contains all the remaining CFFDRS R code translation, along with all tests ported over from the R repo. Tried to maintain all the same comments and references to the source formulas just like the R library.
The review might seem a bit daunting but hopefully the running and passing of all the tests helps. Please see the PR description for additional info on the testing

File: cffdrs/tests/test_distance_at_time.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        fuel_type = row["FUELTYPE"]
+        roseq = row["ROSeq"]
+        hr = row["HR"]
+        cfb = row["CFB"]
+
+        expected_distance = row["DistanceAtTime"]
+
+        calculated_distance = distance_at_time(fuel_type, roseq, hr, cfb)
+
+        assert pytest.approx(expected_distance, abs=0.1, nan_ok=True) == calculated_distance, (
```
Comment (BadgerOnABike) [inline cffdrs/tests/test_distance_at_time.py:28]: In the other tests we are looking at a 1% percent tolerance, here it is 10, is that a typo or was that intentional.

File: cffdrs/tests/test_fbp.py (hunk 1 above)
Comment (BadgerOnABike) [reply cffdrs/tests/test_fbp.py:183]: It is strange as it sets the CFL in the input definition but then re-establishes later on based on the fuel type? So a C-2 would have a CFL of .8 once the crown_fuel_load function runs. It will run if CFL is between 0 and 2, but having a fueltype defined would overwrite? Strange behaviour imo.

File: cffdrs/tests/test_distance_at_time.py (hunk 6 above)
Comment (BadgerOnABike) [reply cffdrs/tests/test_distance_at_time.py:28]: Nevermind, looked up the docs that's an absolute difference not a relative. .1 is probably fine.

File: cffdrs/tests/test_pros.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            row["Long3"],
+            row["Lat3"],
+        )
+        ros_calculated = result.ros
+        direction_calculated = result.direction
+
+        assert pytest.approx(expected_ros, abs=0.01) == ros_calculated, (
+            f"Failed ROS for row {idx}: {row}. Result: {ros_calculated}, Expected: {expected_ros}"
+        )
+        assert pytest.approx(expected_direction, rel=0.1) == direction_calculated, (
```
Comment (BadgerOnABike) [inline cffdrs/tests/test_pros.py]: Is there a reason for using the relative difference? That would allow for a 27 degree difference and I'm thinking that is a pretty big swing.

File: cffdrs/tests/test_sdmc.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+
+@pytest.mark.skip(
+    reason="""Something odd happening with R sdmc tests. R test passes if full test suite is run, but
+            when I run manually and debug I see the same problem that this implements. Needs further investigation.
+                Row 76: Calculated SDMC=0.0484, Expected=0.0484, Diff=-0.0000
+                Row 77: Calculated SDMC=0.0484, Expected=0.0484, Diff=-0.0000
+                Row 78: Calculated SDMC=5.9601, Expected=0.0484, Diff=5.9117
+                Row 79: Calculated SDMC=19.1639, Expected=3.0500, Diff=16.1139
+                Row 80: Calculated SDMC=32.3677, Expected=9.7560, Diff=22.6117"""
```
Comment (BadgerOnABike) [inline cffdrs/tests/test_sdmc.py:25]: Hmm very odd

File: cffdrs/tests/test_pros.py (hunk 7 above)
Comment (brettedw) [reply cffdrs/tests/test_pros.py]: Good catch, I thought I used `abs` errors everywhere.
Looks like if I switch it to `abs` I'm getting a couple errors that are right on that border:

```
E           AssertionError: Failed Direction for row 4: {'Ros': 1.022, 'Direction': -37.67, 'T1': 25.0, 'Long1': -79.701017, 'Lat1': 43.808873, 'T2': 40.0, 'Long2': -79.69966, 'Lat2': 43.809806, 'T3': 75.0, 'Long3': -79.700387, 'Lat3': 43.809806}. Result: -37.77047678091748, Expected: -37.67
E           assert -37.67 ± 0.1 == -37.77047678091748
E             
E             comparison failed
E             Obtained: -37.77047678091748
E             Expected: -37.67 ± 0.1
```

What would you recommend? Would an absolute error of 0.15 suffice?
You're right, this definitely shouldn't be a relative error for direction

File: cffdrs/constants.py (hunk 3 above)
Comment (BadgerOnABike) [reply cffdrs/constants.py:30]: Could also put in Table 7 (FCFDG 1992) for folks reference.

File: cffdrs/direction.py (hunk 5 above)
Comment (BadgerOnABike) [reply cffdrs/direction.py:39]: @jordan-evens I think this would be more in your wheelhouse to discuss than mine, do you have anything to add here?

File: cffdrs/tests/test_pros.py (hunk 7 above)
Comment (brettedw) [reply cffdrs/tests/test_pros.py]: Maybe I'll look at the underlying functions first and see if any improvements can be made

Comment (brettedw) [reply cffdrs/tests/test_pros.py]: Doesn't seem like it, I assume this is likely a small error from R vs Python differences. Even an error tolerance of `abs=0.11` passes

**Comment by BadgerOnABike** [comment]
A few comments and I reopened one of the comments to get the table in the comment.

Otherwise this is a huge lift well done folks, apologies for the delays!

File: cffdrs/tests/test_pros.py (hunk 7 above)
Comment (brettedw) [reply cffdrs/tests/test_pros.py]: Changed error threshold to `abs=0.15`. I imagine for this use case a 0.15 degree absolute error is ok?

Comment (BadgerOnABike) [reply cffdrs/tests/test_pros.py]: That's likely fine, it feels big, but would need to compound the errors along a growth simulation to really understand the magnitude of the issue.

File: cffdrs/tests/test_sdmc.py (hunk 8 above)
Comment (brettedw) [reply cffdrs/tests/test_sdmc.py:25]: Maybe there's something I don't quite understand about how the R package is loading that data for tests

**Comment by BadgerOnABike** [comment]
Worth keeping the discussion going to see if we need to adjust anything but I'd like to unblock further progress on this.
